### PR TITLE
UP my solution

### PR DESCRIPTION
**Pull request by h2o64**

### CODE REVIEW

File: sklearn_questions.py [hunk 1]
```diff
@@ -134,6 +170,21 @@ class MonthlySplit(BaseCrossValidator):
...
+        ----------
+        x : array-like of shape (n_samples, )
+            Training data, where `n_samples` is the number of samples.
+
+        Returns
+        -------
+        n_splits : int
+            The number of splits.
+        """
+        return (int((x.max() - x.min()) / np.timedelta64(1, 'M')) // 2) * 2
```
Comment (tomMoral) [inline sklearn_questions.py:186]: I think this would only work for odd month number no?

Comment (h2o64) [reply sklearn_questions.py:186]: I'll check, thank you !

Comment (h2o64) [reply sklearn_questions.py:186]: @tomMoral it's indeed broken (when I change the end date to "2020-12-31" in the test cases it fails).
I would be a good new test case

File: sklearn_questions.py [hunk 2]
```diff
@@ -152,7 +206,19 @@ def get_n_splits(self, X, y=None, groups=None):
...
+            if not pd.api.types.is_datetime64_any_dtype(X.index):
+                raise ValueError('{} is not datetime compatible'
+                                 .format(self.time_col))
+            return self.get_n_splits_1d(X.index)
+        else:
+            # Check if it is datetime compatible
+            if not pd.api.types.is_datetime64_any_dtype(X[self.time_col]):
+                raise ValueError('{} is not datetime compatible'
+                                 .format(self.time_col))
+            return self.get_n_splits_1d(X[self.time_col])
```
Comment (tomMoral) [inline sklearn_questions.py:221]: To avoid code duplication, you could create a variable that stores the seres you want to work on.
That would make the code more readable and better factorized.

File: sklearn_questions.py [hunk 3]
```diff
@@ -134,6 +170,24 @@ class MonthlySplit(BaseCrossValidator):
...
+
+        Returns
+        -------
+        n_splits : int
+            The number of splits.
+        """
+        max_date = x.max()
+        min_date = x.min()
+        return ((max_date.year - min_date.year) * 12) \
+            + max_date.month - min_date.month
```
Comment (tomMoral) [inline sklearn_questions.py:189]: Yes better!

File: sklearn_questions.py [hunk 4]
```diff
@@ -174,12 +240,24 @@ def split(self, X, y, groups=None):
...
         """
-
-        n_samples = X.shape[0]
+        # Compute the number of splits to do
         n_splits = self.get_n_splits(X, y, groups)
+        # Count number of rows in each month
+        if self.time_col == 'index':
+            X_count_month = X.resample("M").count()
+        else:
+            X_count_month = X.set_index(self.time_col).resample("M").count()
```
Comment (tomMoral) [inline sklearn_questions.py:249]: I think this works only if the data is sorted no?

If you want, could you propose 2 more test cases?
One  with different number of months (related to your previous change) and one with unordered data?

If you add these test case + 1 one more on your own, I think we can give you an extra bonus point.

Comment (h2o64) [reply sklearn_questions.py:249]: I made a PR for those 3 extra test cases here : https://github.com/x-datascience-datacamp/2021-assignment-sklearn/pull/2

Regarding the actual issue, I'm tempted to just sort the data (wrt `time_col`) and call it a day. I'll try to find a better implementation by the end of the class.

Comment (tomMoral) [reply sklearn_questions.py:249]: Yes I think sorting is good enough for the implementation.
For the 3rd test case, maybe you can find a edge case when the data is not continuous. what if we are missing one month?
